File: src/visual/utils/pdf_viewer.py
# Copyright (c) 2022-2026 MKM Research Labs.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from config.theme import colour
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Shared PDF viewer panel factory.

Generates JavaScript for an inline overlay panel that displays
base64-encoded PDFs in an iframe.  Used by PropertyPDFPanel,
GaugePDFPanel, and the insurance claim-report viewer.

The JS body lives in ``src/static/js/pdf_viewer.js`` (plus the optional
custom-event listener in ``pdf_viewer_event.js``); the parameters are spliced
in here via ``__SENTINEL__`` placeholders.
"""


def pdf_viewer_js(namespace: str,
                  panel_width: str = "850px",
                  panel_height: str = "90vh",
                  default_title: str = "Report",
                  btn_color: str = colour('bootstrap-green'),
                  event_name: str = None,
                  event_id_key: str = "id",
                  display_name_js: str = None) -> str:
    """Return a self-contained JS IIFE that creates a PDF viewer popup.

    Args:
        namespace: Unique prefix for DOM ids and the ``window`` API.
            E.g. ``"property-pdf"`` produces ``window.PropertyPDFPanel``.
        panel_width / panel_height: CSS dimensions for the overlay.
        default_title: Fallback title shown in the header.
        btn_color: CSS colour for the Download button.
        event_name: Optional custom-event name to listen for
            (e.g. ``"propertyPdfReady"``).  When fired with
            ``detail.{event_id_key}`` and ``detail.pdfBase64`` the
            panel opens automatically.
        event_id_key: Key inside ``event.detail`` that carries the
            entity id (e.g. ``"propertyId"``, ``"gaugeId"``).
        display_name_js: Optional JS expression that converts ``entityId``
            into a display label (e.g.
            ``"window.propertyDisplayName(entityId)"``).  Falls back to
            bare ``entityId`` if omitted or the expression returns falsy.

    Returns:
        JS string (no ``<script>`` tags — caller wraps if needed).
    """
    # Convert "property-pdf" → "PropertyPDFPanel" for the window API name.
    # Preserves the uppercase "PDF" that consumers expect.
    api_name = "".join(
        seg.upper() if seg == "pdf" else seg.capitalize()
        for seg in namespace.split("-")
    ) + "Panel"

    # Lazy import: pdf_viewer lives in visual.utils, but the js_static helper
    # is inside the visual.interactivity package whose __init__ imports the PDF
    # panels (which import this module). Importing at call time avoids that
    # module-load cycle.
    from visual.interactivity._jsbundle import js_static

    event_listener = ""
    if event_name:
        event_listener = (
            js_static('pdf_viewer_event.js')
            .replace('__EVENT_NAME__', event_name)
            .replace('__EVENT_ID_KEY__', event_id_key)
        )

    return (
        js_static('pdf_viewer.js')
        .replace('__PANEL_W__', panel_width)
        .replace('__PANEL_H__', panel_height)
        .replace('__DEFAULT_TITLE__', default_title)
        .replace('__BTN_COLOR__', btn_color)
        .replace('__DISPLAY_LABEL__', display_name_js or 'entityId')
        .replace('__API_NAME__', api_name)
        .replace('__EVENT_LISTENER__', event_listener)
        .replace('__NS__', namespace)
    )

File: src/static/js/pdf_viewer.js
// Copyright (c) 2022-2026 MKM Research Labs.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

            (function() {
                var PANEL_W = '__PANEL_W__';
                var PANEL_H = '__PANEL_H__';
                var pdfPanel = null;

                function createPanel() {
                    if (pdfPanel) return pdfPanel;

                    pdfPanel = document.createElement('div');
                    pdfPanel.id = '__NS__-panel';
                    pdfPanel.style.cssText =
                        'position:fixed;top:50%;left:50%;transform:translate(-50%,-50%);' +
                        'width:' + PANEL_W + ';height:' + PANEL_H + ';' +
                        'background:var(--panel);border:1px solid var(--divider);border-radius:var(--radius-lg);' +
                        'box-shadow:var(--shadow-toast);z-index:2000;' +
                        'display:none;flex-direction:column;font-family:Arial,sans-serif;';

                    var header = document.createElement('div');
                    header.style.cssText =
                        'display:flex;justify-content:space-between;align-items:center;' +
                        'padding:var(--space-5) var(--space-8);border-bottom:1px solid var(--line-soft);background:var(--wash);' +
                        'border-radius:var(--radius-lg) var(--radius-lg) 0 0;min-height:40px;';

                    var title = document.createElement('span');
                    title.id = '__NS__-title';
                    title.style.cssText = 'font-weight:bold;font-size:var(--size-14);color:var(--text);';
                    title.textContent = '__DEFAULT_TITLE__';

                    var btnGroup = document.createElement('div');
                    btnGroup.style.cssText = 'display:flex;gap:var(--space-4);';

                    var downloadBtn = document.createElement('button');
                    downloadBtn.id = '__NS__-download';
                    downloadBtn.textContent = 'Download';
                    downloadBtn.style.cssText =
                        'padding:var(--space-2) var(--space-6);border:1px solid __BTN_COLOR__;border-radius:var(--radius-4);' +
                        'background:__BTN_COLOR__;color:var(--inverse);cursor:pointer;font-size:var(--size-sm);';
                    downloadBtn.onclick = function() {
                        var iframe = document.getElementById('__NS__-iframe');
                        if (iframe && iframe.src) {
                            var a = document.createElement('a');
                            a.href = iframe.src;
                            a.download = (title.textContent || '__DEFAULT_TITLE__') + '.pdf';
                            a.click();
                        }
                    };

                    var closeBtn = document.createElement('button');
                    closeBtn.textContent = 'Close';
                    closeBtn.style.cssText =
                        'padding:var(--space-2) var(--space-6);border:1px solid var(--red);border-radius:var(--radius-4);' +
                        'background:var(--red);color:var(--inverse);cursor:pointer;font-size:var(--size-sm);';
                    closeBtn.onclick = function() { hidePanel(); };

                    btnGroup.appendChild(downloadBtn);
                    btnGroup.appendChild(closeBtn);
                    header.appendChild(title);
                    header.appendChild(btnGroup);

                    var container = document.createElement('div');
                    container.id = '__NS__-container';
                    container.style.cssText = 'flex:1;overflow:hidden;border-radius:0 0 var(--radius-lg) var(--radius-lg);';

                    var iframe = document.createElement('iframe');
                    iframe.id = '__NS__-iframe';
                    iframe.style.cssText = 'width:100%;height:100%;border:none;';
                    container.appendChild(iframe);

                    pdfPanel.appendChild(header);
                    pdfPanel.appendChild(container);
                    document.body.appendChild(pdfPanel);

                    document.addEventListener('keydown', function(e) {
                        if (e.key === 'Escape' && pdfPanel.style.display !== 'none') {
                            hidePanel();
                        }
                    });

                    return pdfPanel;
                }

                function showPanel(entityId, pdfBase64) {
                    var panel = createPanel();
                    var title = document.getElementById('__NS__-title');
                    var displayLabel = __DISPLAY_LABEL__;
                    title.textContent = '__DEFAULT_TITLE__: ' + (displayLabel || entityId);

                    var iframe = document.getElementById('__NS__-iframe');
                    iframe.src = 'data:application/pdf;base64,' + pdfBase64;

                    panel.style.display = 'flex';
                }

                function hidePanel() {
                    if (pdfPanel) {
                        pdfPanel.style.display = 'none';
                        var iframe = document.getElementById('__NS__-iframe');
                        if (iframe) iframe.src = '';
                    }
                }

                window.__API_NAME__ = {
                    show: showPanel,
                    hide: hidePanel
                };
                __EVENT_LISTENER__
            })();
